### PR TITLE
feat(UI): import OSADL obligation information and update screen of Ad…

### PR DESCRIPTION
**Pull request by quan-tv**
Signed-off-by: Tran Vu Quan <quan1.tranvu@toshiba.co.jp>

> Please provide a summary of your changes here.
> * Import Obligation information from https://www.osadl.org/
> * New GUI for Adding obligation screen

Issue: https://github.com/eclipse/sw360/issues/513

> Have you implemented any additional tests? No

### Checklist
Must:
- [x] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Comment by KoukiHama** [comment]
Due to this https://github.com/eclipse/sw360/issues/1373
Test was not able to pass....

**Comment by KoukiHama** [comment]
I changed CouchDB and Therift version  and solved issue #1373/
So,I think, this pull request error is not related to #1373

**Comment by ravi110336** [comment]
Hi @quan-tv , I tried to give a clean mvn build with the test cases and for me it failed in the resource-server package. I have tried with thrift 0.13 version.

image

Error log:
image

**Comment by ravi110336** [comment]
Getting the same error as above with the thrift 0.14 version also.

**Comment by ravi110336** [comment]
Getting the same error as above with the thrift 0.14 version also.

**Comment by KoukiHama** [comment]
Me too. 
image


And this is log about error.
[error_osdal_build.txt](https://github.com/eclipse/sw360/files/7176453/error_osdal_build.txt)

**Comment by KoukiHama** [comment]
I could build success  cdf4ec3

**Comment by ravi110336** [comment]
Hi @KoukiHama , The build is success.

**Comment by KoukiHama** [comment]
thank you @ravi110336 ,

but I was not able to find the reason of GitHub action error yet. 

@quan-tv  have you already got some clues about GitHub action error?

[error_debug_option.txt](https://github.com/eclipse/sw360/files/7200474/error_debug_option.txt)

This logs comes from "mvn install" with debug option corresponding to GitHub Action's error.

**Comment by quan-tv** [comment]
@KoukiHama 

Sorry, I got no idea what is the cause of this failure.

**Comment by KoukiHama** [comment]
I got full log of "mvn package" with "-X" debug option. 
(I used https://github.com/nektos/act for this)
[log.txt.zip](https://github.com/eclipse/sw360/files/7205351/log.txt.zip)

**Comment by quan-tv** [comment]
@KoukiHama 
Could you please tell me how to with '-X' option?

> mvn install -X

Is this correct?

**Comment by KoukiHama** [comment]
@quan-tv 

> mvn install -X
> Is this correct?

YES. I used -X option.

**Comment by KoukiHama** [comment]
My local environment use open_jdk
Github runner use Adopt_jdk

Is this related to this problem_

**Comment by KoukiHama** [comment]
(from OSDAL members in sw360 telco )
Add OSDAL info and CC-BY-4.0 into Code

**Comment by quan-tv** [comment]
Dear @KoukiHama -san, @ravi110336 -san,

The source code was updated and can be built successfully now.

Note: OSADL info and CC-BY-4.0 were not added yet.

File: backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/LicenseHandler.java [hunk 1]
```diff
@@ -265,6 +313,14 @@ public RequestSummary importAllSpdxLicenses(User user) throws TException {
         return handler.importAllSpdxLicenses(user);
     }
 
+    @Override
+    public RequestSummary importAllOSADLLicenses(User user) throws TException {
+        if(! PermissionUtils.isUserAtLeast(UserGroup.CLEARING_ADMIN, user)){
```
Comment (ravi110336) [inline backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/LicenseHandler.java:318]: Indentation after if and before ( same before the {, please look into the code for similar indentation issue,most of the indentation issue i have highlighted.

File: backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/db/LicenseDatabaseHandler.java [hunk 2]
```diff
@@ -206,6 +218,49 @@ public String addObligations(@NotNull Obligation obligs, User user) throws SW360
         return obligs.getId();
     }
 
+    /**
+     * Adds a new obligation element to the database.
+     *
+     * @return ID of the added obligations element.
+     */
+    public String addObligationElements(@NotNull ObligationElement obligationElement, User user) throws SW360Exception {
+        if (!PermissionUtils.isUserAtLeast(UserGroup.CLEARING_ADMIN, user)){
```
Comment (ravi110336) [inline backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/db/LicenseDatabaseHandler.java:227]: Indentation before { and in line 249,714.

File: backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/db/LicenseDatabaseHandler.java [hunk 3]
```diff
@@ -762,4 +952,106 @@ public int checkLicenseTypeInUse(String id) {
...
+
+    // Read nodes from input and save
+    public String addNodes(String jsonString, User user) throws SW360Exception {
+        try {
+            com.liferay.portal.kernel.json.JSONObject jsonObject = JSONFactoryUtil.createJSONObject(jsonString);
+            return addNodes(jsonObject, user);
+        }
+        catch (Exception e){
+            //TODO: handle exception
+        }
```
Comment (ravi110336) [inline backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/db/LicenseDatabaseHandler.java:964]: may be we can handle the exception here and in the below catch blocks as well.
please also check the indentation here in the catch block, and in below catch block before { .
indention in line 993 after if. line 1031 after for.

File: backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/tools/OSADLObligationConnector.java [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+			try {
+				JSONObject currentLine = JSONFactoryUtil.createJSONObject(line);
+				int parentLineId = getParentLineId(currentLine, lines);
+				if (parentLineId >= 0) {
+					try {
+						JSONObject parentLine = JSONFactoryUtil.createJSONObject(lines.get(parentLineId));
+						parentLinePath = parentLine.get("path").toString();
+					} catch (Exception e) {
+						//TODO: handle exception
+					}
```
Comment (ravi110336) [inline backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/tools/OSADLObligationConnector.java:132]: Please handle the exception in the catch block above and in the below catch blocks.
Indentation in line 208 before {

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java [hunk 5]
```diff
@@ -379,6 +383,12 @@
     public static final String LICENSE_INFO_EMPTY_FILE = "isEmptyFile";
     public static final String SW360_USER = "sw360User";
 
+    //! Specialized keys for obligation node
+    public static final String OBLIGATION_NODE_LIST = "obligationNodeList";
+
+    //! Specialized keys for obligation element
+        public static final String OBLIGATION_ELEMENT_LIST = "obligationElementList";
```
Comment (ravi110336) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java:390]: alignment issue, please remove some space.

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/admin/licenseAdmin/view.jsp [hunk 6]
```diff
@@ -137,11 +152,75 @@ require(['jquery', 'modules/dialog', 'modules/validation'], function($, dialog,
         );
     });
 
+    $('.portlet-toolbar button[data-action="import-OSADL"]').on('click', function() {
+        var $dialog;
+        if(progress != null) {
```
Comment (ravi110336) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/admin/licenseAdmin/view.jsp:157]: Indentation after if and same for line 218, in line 272 before { .

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/admin/obligations/obligationTextTree.jsp [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+        });
+
+        const typeSuggestions = getTypeSuggestions();
+
+        const obligationSuggestions = getObligationSuggestions();
+
+        function getJSONObjectKeys(object) {
+            var keys = [];
+
+            for(var key in object) {
```
Comment (ravi110336) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/admin/obligations/obligationTextTree.jsp:177]: indentation after for and in line 422 after if .

**Review by ravi110336 — COMMENTED** [review]
Code looks good to me please consider some minor review comments.

**Comment by ravi110336** [comment]
Could you please guide me how to get the obligation information from the below link?

- Import Obligation information from https://www.osadl.org/

**Comment by ravi110336** [comment]
Also, could you please check if it is possible to squash these 4 commits into one single commit?

**Comment by ravi110336** [comment]
Tested the import functionality after registering a new license to sw360, and it works fine. Could you please check one Observation below and comment. (Please refer below Observation screenshot 2:)

**Screenshot 1:** (Import functionality works fine).
image

**Observation:** I tried to add a new obligation from the Admin->Obligations-> Add obligation
In the language element dropdown it shows **"YOU NOT MUST"** but i think the text should be **"YOU MUST NOT"**. Could you please comment on this behavior if this ok for the license text, because as when we import the OSADL information the text is like **"YOU MUST NOT Promote"**
**Screenshot 2:**
image
image

**Comment by ravi110336** [comment]
As per the discussion in the call need to test the CLI obligations.

**Comment by ravi110336** [comment]
Tested the CLI obligations and it is working fine.

**Pull request merged**